### ISSUE TITLE
injectable logger for `TopKSAETrainer`

### ISSUE THREAD

**Issue by zombak79** [zombak79/compresso#16]
# Spec: injectable logger for `TopKSAETrainer`

**Repo:** compresso (`compresso-pytorch`) · **Affects:** `src/compresso/trainers/saetrainer.py`

## Problem

`TopKSAETrainer.fit()` is observable only two ways today, and neither works for a long-running
containerised job:

1. **tqdm** (`show_progress=True`). Writes a progress bar to stderr. In a container there is no
   tty, so every refresh becomes its own log line, interleaved with the host application's own
   structured logging. `_progress()` also hardcodes `tqdm(iterable, total=total)` with no `file=`
   parameter, so the destination cannot be redirected by the caller.
2. **`trainer.history`** after `fit()` returns. Complete, but only available at the end. A
   30-epoch run is silent for its whole duration, which makes a slow epoch indistinguishable
   from a hung process.

There is no third option: `fit()` accepts no callback, and its only per-epoch hooks are
`dataset.on_epoch_begin()` / `on_epoch_end()`, which fire *before* the epoch record (including
`val_loss`) is appended to `history` at line 992. Reaching them also requires overriding the
private `_dataset()`.

Separately, the tqdm postfix carries only `loss`, `cosine`, `mse`, `lr` and `val_loss`. It omits
**`dead_features`**, which is the diagnostic that says whether `hidden_dim` is too large for the
catalog — the main tuning question when adapting a small item set.

**Current workaround in python-models:** a watcher thread polling `trainer.history`
(`embedding_adaptors/SAE.py::_stream_epochs`). It works and touches no compresso internals, but
it is a thread per training run to compensate for a missing parameter, and it can only report
what lands in `history`. This spec is what would let us delete it.

## Proposed API

Two config fields and one constructor argument. All default to today's behaviour.

```python
class TopKSAETrainer:
    def __init__(self, config: TopKSAEConfig | None = None, logger: Any | None = None) -> None:
```

`logger` is any object exposing `.info(str)` — `logging.Logger`, python-models' `RemoteLogger`,
or a bare function wrapper. Deliberately duck-typed: compresso must not gain a logging
dependency, and callers already have their own logger.

```python
@dataclass(frozen=True)
class TopKSAEConfig:
    ...
    log_prefix: str = "TopKSAE"      # prepended to every line, so the output is greppable
    log_every_n_steps: int = 0       # 0 = per-epoch only; N>0 also logs every N batches
```

### Behaviour

| `logger` | `show_progress` | Result |
|---|---|---|
| `None` | `True` | tqdm, exactly as today (**unchanged default**) |
| `None` | `False` | silent, exactly as today |
| provided | either | structured lines via `logger.info`; tqdm suppressed regardless, since the two would duplicate |

Suppressing tqdm when a logger is given is the one behaviour change, and it only affects callers
who opt in by passing `logger`.

### Lines to emit

**On fit start** — one line: input dim, `hidden_dim`, `k`, rows (train/validation split),
batches per epoch, epochs, `patience`, device, and whether corruption or standard scaling is
active. This is what makes a log self-describing six months later.

**On each epoch end** — one line carrying **every key in the epoch's `history` record**, not a
curated subset. That is the requirement that matters: it must include `dead_features` and
`active_count`, and must keep including any key added to `history` later without another change
here. Plus epoch number, elapsed, and an ETA for the remaining epochs.

**On each Nth batch**, when `log_every_n_steps > 0` — step number of total, time per step,
elapsed, remaining. Useful when a single epoch is minutes long.

**On fit end** — total duration, `epochs_run`, `best_epoch`, `best_val_loss`, and whether early
stopping fired (`stopped_epoch`).

### Formatting

Follow `util/sparse_elsa/callbacks.py::loggerCallback` in python-models, which is the
established shape for this:

- `[{log_prefix}] ` prefix on every line.
- `%.4f` for values with `abs(v) > 1e-3`, `%.4e` below — a `1e-7` loss must not print as `0.0000`.
- Durations auto-scaled to s / ms / µs.
- ETA from mean time per epoch so far.

## Constraints

- **No behaviour change when `logger` is omitted.** Existing callers, including the notebooks,
  must be unaffected.
- **No new dependencies.** tqdm is already an optional soft import; keep it that way.
- **Never let logging break training.** A raising or misbehaving logger must not abort a
  multi-hour fit — wrap emission so a logging failure is swallowed (or reported once and then
  disabled).
- **Cheap.** Per-epoch is free. Per-batch at `log_every_n_steps=1` should not dominate a small
  model's step time; the guard should be an integer modulo before any string formatting.

## Alternative considered: a callback protocol

A general `callbacks: list[TrainerCallback]` with `on_train_begin` / `on_epoch_end` / … is more
flexible — it would also serve metrics export and custom early stopping. It is not what we are
asking for, because compresso already owns early stopping and LR scheduling internally, so the
only present need is observability, and a callback protocol is a larger API surface to design
and keep stable. If callbacks arrive later, a `logger` argument is trivially reimplementable as
a built-in callback, so this does not paint anyone into a corner.

## Consumer change once this lands

In `embedding_adaptors/SAE.py`: delete `_stream_epochs`, `_format_epoch`, the
`HISTORY_POLL_SECONDS` constant and the thread plumbing in `fit()`; construct the trainer as
`TopKSAETrainer(self._build_compresso_config(), logger=self.dao.logger)` and set
`log_prefix="SAE"`. Keep `show_progress=False`. Net effect in python-models is a deletion.

## Acceptance

- Default construction (`logger=None`) produces byte-identical output to the current version,
  with `show_progress` both `True` and `False`.
- With a logger, a 3-epoch fit emits: 1 start line, 3 epoch lines, 1 end line — and each epoch
  line contains `dead_features`, `active_count` and `val_loss`.
- Adding a new key to a `history` record makes it appear in the epoch line with no further edit.
- A logger whose `.info` raises does not fail the fit.
- `log_every_n_steps=1` on a 3-batch epoch emits 3 step lines per epoch.
